clients/java/src/main/java/io/camunda/zeebe/client/impl/http/CircularBufferInputStream.java

```diff
@@ -123,39 +128,58 @@ public synchronized int read(@Nonnull final byte[] b, final int off, final int l
       return 0;
     }
 
-    while (available == 0) {
+    if (canConsumeData()) {
+      return -1;
+    }
+
+    int bytesRead = 0;
+    final int bytesToRead = Math.min(len, availableData);
+
+    final int firstCopyLength = Math.min(bytesToRead, capacity - readPos);
+    System.arraycopy(buffer, readPos, b, off, firstCopyLength);
+    readPos = (readPos + firstCopyLength) % capacity;
+    bytesRead += firstCopyLength;
+    availableData -= firstCopyLength;
+
+    final int remaining = bytesToRead - firstCopyLength;
+    if (remaining > 0) {
+      System.arraycopy(buffer, readPos, b, off + firstCopyLength, remaining);
+      readPos = (readPos + remaining) % capacity;
+      bytesRead += remaining;
+      availableData -= remaining;
+    }
+
+    // Notify any waiting writers
+    notifyAll();
+
+    // Inform capacity callback
+    if (capacityCallback != null) {
+      capacityCallback.onCapacityAvailable(bytesRead);
+    }
+
+    return bytesRead;
+  }
+
+  private boolean canConsumeData() throws IOException {
```

Comment:
🔧 This can be refactored as follow if I'm not wrong:

```
private boolean canConsumeData() throws IOException {
    while (availableData == 0 && exception == null && !endOfStream) {
        try {
            wait();
        } catch (final InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new IOException("Interrupted while reading", e);
        }
    }

    if (exception != null) {
        IOException ex = exception;
        exception = null; 
        throw ex;
    }
    return availableData == 0 && endOfStream;
}
```
Adding also `exception = null; ` to avoid repeated exceptions being thrown

Reply:
Also the `endOfStream` check in the last line seems redundant as we are already checking it in the first condition